### PR TITLE
Add a garbage collector that deletes entities outside of level bounds…

### PR DESCRIPTION
**Pull request by FabioGNR**
… (excluding players, which are marked 'essential')

### CODE REVIEW

File: game/systems/GarbageCollectorSystem.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+            , m_camera{ camera }
+        {
+        }
+
+        void update(std::chrono::nanoseconds timeStep) override;
+        void render(engine::IRenderer& renderer) override;
+
+    private:
+        engine::ecs::World& m_ecsWorld;
+        std::shared_ptr<engine::graphics::Camera> m_camera;
```
Comment (notzain) [inline game/systems/GarbageCollectorSystem.h]: Discussed in expert input: be careful where you use shared_ptr.

unique_ptr and shared_ptr are for _ownership_, meaning that the GCSystem now owns the camera. This is probably wrong. Use a raw pointer (or ref_wrapper if you like increasing your character count) instead.

https://stackoverflow.com/questions/8338570/c11-replace-all-non-owning-raw-pointers-with-stdshared-ptr

File: game/systems/GarbageCollectorSystem.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    common::Vector2D<double> levelBounds{ 0, 0 };
+    m_ecsWorld.forEachEntityWith<components::LevelMetaComponent, components::DimensionComponent>([&](engine::ecs::Entity& levelEntity) {
+        levelBounds = m_ecsWorld.getComponent<components::DimensionComponent>(levelEntity).dimension;
+    });
+    std::vector<engine::ecs::Entity*> entitiesToCollect;
+    m_ecsWorld.forEachEntityWith<components::PositionComponent, components::DimensionComponent>([&](engine::ecs::Entity& entity) {
+        if (entity.hasComponent<components::EssentialComponent>()) {
+            return; // this entity may not be collected automatically
+        }
+        auto position = m_ecsWorld.getComponent<components::PositionComponent>(entity).position;
```
Comment (notzain) [inline game/systems/GarbageCollectorSystem.cpp]: const auto& -> you dont modify the code, nor need a copy

**Comment by FabioGNR** [comment]
Entities are deleted if they are not visible (the camera check) AND if they are outside the level bounds. This means tiles will not be deleted because they do not move outside the level bounds. The player can be removed when out of bounds but I think for the score system etc it's easier to keep them somewhere. Or store that info somewhere else, but then that should be implemented separately.

The shared ptr is because both the camerasystem and the rendering system depend on the camera. So I thought they should own it. Then because this collector also requires the camera to function I gave it the shared pointer as well.

Is it better to let the GameState own the camera?

I will fix the const auto&

**Comment by notzain** [comment]
I still feel like tagging entities as destructable is clearer. Case in point: you now have exceptions for the objective of the GCSystem.

> "The GarbageCollectorSystem deletes all objects that have a Position and Dimension component, _except_ those that are tagged as essential (or any other reason we might think of in the future)"

vs.

> "The GarbageCollector deletes all objects tagged as Destructable"

This also allows us to seperate concerns: the GCSystem deletes all Destructables. Then we can add other systems, that tag entities as Destructable, such as the out of level boundary check.

**Comment by notzain** [comment]
> The shared ptr is because both the camerasystem and the rendering system depend on the camera. So I thought they should own it. Then because this collector also requires the camera to function I gave it the shared pointer as well.

They both **use** the camera, but they don't own it. This concept might be unique to C++, and is hard to understand and wrap your head around.

Try asking yourself this question: "If class X gets destroyed, should shared_ptr Y also be destroyed?"

`If the GarbageCollectorSystem gets destroyed, should the Camera also be destroyed?` 
Probably not.

`If the RenderingSystem gets destroyed, should the Camera also be destroyed?` 
Seems plausible. Is the camera used anywhere else besides the RenderingSystem?

`If the Game(State) gets destroyed, should the Camera also be destroyed?` 
Yes.

**Comment by FabioGNR** [comment]
> > `If the RenderingSystem gets destroyed, should the Camera also be destroyed?`
> Seems plausible. Is the camera used anywhere else besides the RenderingSystem?
> 
> `If the Game(State) gets destroyed, should the Camera also be destroyed?`
> Yes.

There's also a CameraSystem which moves it. If a rendering system does not exist to use the camera, the camera would lose all purpose. But making the rendering system the owner will be difficult because other systems also need it. So I guess it would be best to make the GameState the owner and then pass references.

**Comment by FabioGNR** [comment]
> I still feel like tagging entities as destructable is clearer. Case in point: you now have exceptions for the objective of the GCSystem.
> 
> > "The GarbageCollectorSystem deletes all objects that have a Position and Dimension component, _except_ those that are tagged as essential (or any other reason we might think of in the future)"
> 
> vs.
> 
> > "The GarbageCollector deletes all objects tagged as Destructable"
> 
> This also allows us to seperate concerns: the GCSystem deletes all Destructables. Then we can add other systems, that tag entities as Destructable, such as the out of level boundary check.
> 
> I will leave this open for the 2nd reviewer to decide.

Separating the concerns seems to be a somewhat unnecessary step to me.
The actual 'deleting/collecting' of garbage is one call to the ECS world `destroyEntity`.
If you separate the out of level boundary check into a new system, this new system will basically do the same thing this garbage collector is currently doing. Except it would add the Destructable component instead of deleting it directly. But there might still be a reason to exclude certain entities from deletion (currently players, though this might change). Maybe I don't fully grasp what you're suggesting, I'm open to changing it but not sure what is the right solution. Hopefully a second reviewer can help with it :)

Thanks for the feedback so far.

**Comment by oblekta** [comment]
Interesting discussion guys.

I prefer the "The GarbageCollector deletes all objects tagged as Destructible" objective, because it "feels" better to assign an entity as destructible instead of making exceptions. You're in complete control over when you want to remove a certain entity.

The problem is that we have to be careful when the next system wants to use an already deleted entity.  As far as I know we have no NULL checks within the foreach functions of the ECS framework (I could be wrong though). The GarbageCollector definitely needs to delete the entities at the end of each tick (low priority). 

Another "issue" is that we need to create multiple systems for setting a Destructible component. In this case the bullets probably need a BulletDestructibleSystem that should detect when a bullet is out of bounds (and not visible on screen). What will happen when we have different kind of projectiles with different behaviors? Can we create a generic system to handle that behavior or will we have multiple systems to satisfy this.

Anyways, I can't decide for you guys at this point. Maybe time is also a valid reason to stick to this implementation for now. Let's discuss this tomorrow.

**Comment by notzain** [comment]
> The problem is that we have to be careful when the next system wants to use an already deleted entity. As far as I know we have no NULL checks when searching/asking for a component within the forEachComponent method (I could be wrong though). The GarbageCollector definitely needs to delete the entities at the end of each tick (low priority).

This issue isn't null checks, but iterator invalidation. When you are looping an array, and inside the loop delete elements of that array, your index gets messed up.

> Another "issue" is that we need to create multiple systems for setting a Destructible component.

This problem also applies the current implementation. ECS makes it harder to make generic systems, and sometimes you have to duplicate code.

**Comment by FabioGNR** [comment]
I've separated the garbage collector into two systems:

1. GarbageCollector: gets all entities with DestructibleComponent and deletes them
2. ProjectileDestroyer: loops all projectiles and marks them destructible if they are out of bounds and not visible to camera.

I also made some changes to how the camera works so that it tries to limit showing anything that's outside the level (works better with a larger level).

**Pull request merged**